File: Python/src/advanced/multiple_inheritance.py

class VeicoloGenerico:
  __secret="sono un segreto" # incapsulamento
  def muovi(self):
    print("Non raggiungibile, errore")
  @property
  def secret(self):
    return self.__secret
  @secret.setter
  def secret(self, value):
    self.__secret = value
  @secret.deleter
  def secret(self):
    del self.__secret


class Auto(VeicoloGenerico):
  def muovi(self):
    print("Veicolo Auto da ora")
  def guida(self):
    self.muovi()
    print("Guida su strada")


class Barca(VeicoloGenerico):
  def muovi(self):
    print("Veicolo Barca da ora")
  def naviga(self):
    self.muovi()
    print("Naviga sull'acqua")


class AmphibiousCar(Auto, Barca):
  __more_secret="sono un segreto senza getter, variabile di classe" # incapsulamento
  def __init__(self):
    Auto.__init__(self)
    Barca.__init__(self)
    self.__more_secret="""
    sono un segreto senza getter, ora variabile di istanza
    non più variabile di classe senza getter quindi solo
    ora vengo aggiunta al dizionario di classe
    """ # incapsulamento
  def muovi(self):
    print("Veicolo anfibio in movimento da ora")


class TestVeicoli:
  @classmethod
  def run(cls):
    print("--- Test property ---")
    gv=VeicoloGenerico()
    print(gv.secret)
    gv.secret="sono un segreto nuovo"
    print(gv.secret)
    print("--- Test AmphibiousCar ---")
    boat = Barca()
    car = Auto()
    amphibious = AmphibiousCar()
    car.guida()  
    boat.naviga()     
    amphibious.muovi()    
    amphibious.naviga()   
    amphibious.guida()   
    print(amphibious.__dict__['_AmphibiousCar__more_secret']) # accesso come variabile di istanza
    print("--- Test polimorfismo ---")
    veicoli = [Barca(), Auto(), AmphibiousCar()]
    for veicolo in veicoli:
      veicolo.muovi()  # Chiamata polimorfica

File: cpp/src/advanced/multiple_inheritance.cpp
#include "multiple_inheritance.h"


// costruttore e distruttore virtuale non necessari
void VeicoloGenerico::muovi() { cout << "Non raggiungibile, errore" << endl; } // virtual in class declaration only
void Auto::muovi()  { cout << "Veicolo Auto da ora" << endl; } // override must be explicit in class declaration only
void Auto::guida() { muovi(); cout << "Guida su strada" << endl; }
void Barca::muovi()  { cout << "Veicolo Barca da ora" << endl; } 
void Barca::naviga() { muovi(); cout << "Naviga sull'acqua" << endl; }
void AmphibiousCar::muovi() { cout << "Veicolo anfibio in movimento da ora" << endl; }

void advanced_class_example(){

  cout << "--- Test AmphibiousCar ---" << endl;

  Barca boat;
  Auto car;
  AmphibiousCar amphibious;
  car.guida();             // Output: "Guida su strada"
  boat.naviga();           // Output: "Naviga sull'acqua"
  amphibious.muovi();      // Output: "Veicolo anfibio in movimento"
  amphibious.naviga();
  amphibious.guida();

  cout << "--- Test polimorfismo ---" << endl;

  const int VEICOLI=3;
  VeicoloGenerico** vettoreVeicoli = new VeicoloGenerico*[VEICOLI];
  vettoreVeicoli[0] = new Barca();
  vettoreVeicoli[1] = new Auto();
  vettoreVeicoli[2] = new AmphibiousCar();

  // polimorph, si usa il puntatore
  for (int i = 0; i < VEICOLI; i++) { vettoreVeicoli[i]->muovi(); }

  // Deallocazione della memoria
  for (int i = 0; i < VEICOLI; i++) { delete vettoreVeicoli[i]; }
  delete[] vettoreVeicoli;

}
